Add syscall/js

hajimehoshi opened the pull request:
This PR adds a new package `syscall/js`. Originally `syscall/js` is for WebAssembly, but in GopherJS this package works as a wrapper of `github.com/gopherjs/gopherjs/js`. With this package, developers can use `syscall/js` both for GopherJS and WebAssembly.

The implementation is basically same as `github.com/gohperjs/gopherwasm`.

Fixes #899

hajimehoshi commented:
```
--- FAIL: TestSleep (0.10s)
    test.148654449:38: Sleep(100ms) slept for only 99ms
```

Meh, this sounds like a flaky issue

neelance commented:
@hajimehoshi While working on WebAssembly I've noticed that `setTimeout` sometimes invokes the callback 1ms early. Maybe this is the same issue. It would probably be okay to always add 1ms to the delay.

hajimehoshi commented:
@neelance Thank. I think I'll do this in another PR later.

hajimehoshi commented on compiler/natives/fs_vfsdata.go:
```diff
@@ -387,10 +387,10 @@ var FS = func() http.FileSystem {
+			uncompressedSize: 36239,
 
-			compressedContent: []byte("\x1f\x8b\x08\x00\x00\x00\x00\x00\x00\xff\xec\x7d\x7f\x73\xe3\x36\xb2\xe0\xdf\xd2\xa7\xc0\xa8\xb6\xfc\xc8\x19\xae\x6c\x39\x7b\xa9\x94\x33\xce\xab\xcd\x24\xd9\xf3\x6e\x66\x9c\x1a\xef\xe4\xae\xce\xcf\x35\x0b\x8b\xa0\x04\x8b\x02\xb9\x24\xa4\xb1\x62\xfb\xbb\x5f\xa1\x1b\xbf\x49\xca\xf6\x24\xb9\xdb\x7a\xb5\xf3\xc7\x58\x22\x81\x46\xa3\xbb\xd1\xe8\x5f\x80\x0e\x0f\xc9\xab\xeb\ ...
+			compressedContent: []byte("\x1f\x8b\x08\x00\x00\x00\x00\x00\x00\xff\xec\x7d\x7f\x73\xe3\x36\xb2\xe0\xdf\xd2\xa7\xc0\xa8\xb6\xfc\xc8\x19\x2e\x3d\x72\xf6\x52\x29\x67\x9c\x57\x9b\x49\xb2\xe7\xdd\xcc\x38\x15\xef\xe4\xae\xce\xcf\x95\x85\x49\x50\x82\x45\x81\x5c\x12\xd2\x58\xf1\xf8\xbb\x5f\xa1\x1b\xbf\x49\xc9\x9e\x49\x72\xb7\xf5\x6a\xe7\x8f\xb1\x44\x02\x8d\x46\x77\xa3\xd1\xbf\x00\x1d\x1f\x93\x17\x37\ ...
```
Thank you for triaging!

hajimehoshi commented on tests/syscalljs_test.go:
```diff
@@ -0,0 +1,275 @@
+	"testing"
+)
+
+func TestSyscallJsNull(t *testing.T) {
```
Done.

hajimehoshi commented:
@neelance Hi, are you happy with this PR? Thanks!

hajimehoshi commented:
ping @neelance

neelance commented:
Is this passing the normal tests of the `syscall/js` package?https://github.com/golang/go/blob/master/src/syscall/js/js_test.go

Is there any particular need for the custom test cases of this PR?

hajimehoshi commented:
To test `js_test.go`, we'd need to copy the entire file except for `// +build js,wasm`. Should we do that?

hajimehoshi commented:
BTW, I put a test file at `compiler/natives/src/syscall/js`, but the test was not executed, I'm not sure the reason though.

neelance commented:
Shouldn't it be possible to modify `importWithSrcDir` to manually include `js_test.go` in the build?

hajimehoshi commented:
Almost done! I found a problem in `js_test.go`:

```
func TestIntConversion(t *testing.T) {                                                                                                                                                                                                    
        testIntConversion(t, 0)                                                                                                                                                                                                           
        testIntConversion(t, 1)                                                                                                                                                                                                           
        testIntConversion(t, -1)                                                                                                                                                                                                          
        testIntConversion(t, 1<<20)                                                                                                                                                                                                       
        testIntConversion(t, -1<<20)                                                                                                                                                                                                      
        testIntConversion(t, 1<<40)                                                                                                                                                                                                       
        testIntConversion(t, -1<<40)                                                                                                                                                                                                      
        testIntConversion(t, 1<<60)                                                                                                                                                                                                       
        testIntConversion(t, -1<<60)                                                                                                                                                                                                      
}
```

This cannot be compiled on 32bit-integer environment and GopherJS is an environment with 32bit integers. Replacing the test function in a usual GopherJS way didn't work. Probably the const check happens before the replacing. Any suggestions?

neelance commented:
This is unfortunate. Okay, maybe coping `js_test.go` is the simplest solution.

hajimehoshi commented:
OK, I copied the test, but now I'm facing a new error on my machine:

```
$ go generate ./... && go run . test syscall/js
writing fs_vfsdata.go
writing fs_vfsdata.go
gopherjs: Source maps disabled. Install source-map-support module for nice stack traces. See https://github.com/gopherjs/gopherjs#gopherjs-run-gopherjs-test.
testing: warning: no tests to run
PASS
[...]
```

Looks like the file is compiled Now it's midnight, I'll investigate this further later...

hajimehoshi commented:
@neelance In the current implementation, is it possible to 'add' new tests in compiler/native?

hajimehoshi commented:
>  is it possible to 'add' new tests in compiler/native?

I thought it impossible, so I moved `js_test.go` in compiler/native to `tests/syscalljs/js_test.go`. Also I found some tests that cannot pass, so I added `t.Skip`.

* `TestNaN`: comparing `NaN` doesn't work with `*js.Object`
* `TestZeroValue`: the initial value of `*js.Object` is nil, which is not `undefined`.

Please take a look

neelance commented:
The zero value is quite important. I think it should be possible to handle this correctly, shouldn't it?

hajimehoshi commented:
> I think it should be possible to handle this correctly, shouldn't it?

Yes. I fixed that so that the zero value is js.Undefined.

I appreciate your patience and your support. Please take another look.

hajimehoshi commented:
ping

johanbrandhorst commented:
This is fantastic :tada:. Well done @hajimehoshi
